csrc/scheduler/pointwise_utils.cpp

```diff
@@ -142,6 +144,104 @@ bool DomainMap::areAllInputIdsMappedTo(TensorView* input_tv, TensorView* tv)
+// an easy way to check that.
+// Instead of that, this function checks that all source iter domains involved
+// in transformation on output_tv is covered by reference_tv. We do so by
+// traverse all disjoint set producers on both tvs and filter them with
+// `ca_map_.uniqueExactDefinitions(id).empty()`.
+//
+// ------
+//
+// e.g 0.
+//   T34 [i0, i1]
+//   T185 [i0, b2, i1]     = broadcast(T34)
+//   T192 [i0, b3(ex), i1] = expand(T185)
+//   T198 [i0, b3(ex)*i1]  = reshape(T192)
+//   output(T34)
+//   output(T198)
+//
+// if we consider taking T34 as reference_tv. T198 is the output_tv. We can't
+// replay T34's transform of merging all the dimensions to T198, since b3(ex)*i1
+// can't be reversed. The check in this function would give us T34 with source
+// i0, i1; where T198 would have source i0, b3, i1, where b3 isn't contained in
+// T34. Hence we'll reject this referenc_tv.
+//
+// ------
+//
+// e.g 1.
+//   T0 [i0, i1]
+//   T1 [i2, i0, i1]
+//   T2 [i0*i1]      = reshape(T0)
+//   T3 [b3, i0, i1] = broadcast(T0)
+//   T4 [i2, i0, i1] = add(T1, T3)
+//   output(T2)
+//   output(T4)
+//
+// the example above should be able to pick T4 as reference_tv. T2's source i0,
+// i1 are both contained by the source of T4, so this example could be scheduled
+// as a single fusion.
+bool DomainMap::areAllOutputIdsMappedTo(
+    TensorView* output_tv,
+    TensorView* reference_tv) const {
+  // traverse back to collect all disjoint set producers from the logical domain
+  // of tv.
+  auto get_source_producers = [this](TensorView* tv) {
+    VectorOfUniqueEntries<std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>>
+        all_producer_sets;
+    std::for_each(
+        tv->getLogicalDomain().begin(),
+        tv->getLogicalDomain().end(),
+        [&](IterDomain* id) {
+          all_producer_sets.pushBack(
+              ca_map_.disjointSetOf(id, IdMappingMode::EXACT));
+        });
+    all_producer_sets.pushBack(
+        ca_map_.getAllDisjointSetProducers(all_producer_sets));
+
+    std::vector<IterDomain*> source_ids;
+    std::for_each(
+        all_producer_sets.vector().begin(),
+        all_producer_sets.vector().end(),
+        [&source_ids,
+         this](const std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>&
+                   producer_set_ptr) {
+          IterDomain* id = producer_set_ptr->front();
+          if (ca_map_.uniqueExactDefinitions(id).empty()) {
+            source_ids.push_back(id);
+          }
+        });
+    return source_ids;
+  };
+
+  // this contains all source iter domain that's covered by reference_tv, so
+  // it's safe for output_tv to have them.
+  std::unordered_set<IterDomain*> covered_source_ids;
+  for (IterDomain* id : get_source_producers(reference_tv)) {
+    covered_source_ids.insert(id);
+  }
+  // it's safe to have source iter domain on output_tv that's not in
+  // reference_tv, since they are not involved in any transforms.
```

Comment:
I wonder why. Can you show  an example?

Reply:
In this example

```
Segmented_Fusion Dump: -- fusion segments:
Segmented_Fusion{
groups:
  pointwise{1, 2, 3, 7, 8, 28}
edges:

group details:
g{(pointwise)
group id: 0
inputs:
  T0_g_float[bS0{1}, iS1{2048}, iS2{512}] float
  T1_g_float[iS3{7}, iS4{8}, iS5{2048}, iS6{64}] float
outputs:
  T8_g_float[bS37{1}, iS38{8}, bS39{1 ex 4}, iS40{2048}, iS41{64}] float
  T9_g_float[iS42{7}, iS43{8}, iS44{2048}, iS45{64}] float


T2_l_float[bS7{1}, iS8{2048}, iS11{8}rf, iS12{64}rf] = view( T0_g_float[bS0{1}, iS1{2048}, iS2{512}] )
(1)
T3_l_float[bS13{1}, iS15{8}, iS14{2048}, iS16{64}]
   = Set.Permute( T2_l_float[bS7{1}, iS8{2048}, iS11{8}rf, iS12{64}rf], cache_op=Streaming )
(2)
T9_g_float[iS42{7}, iS43{8}, iS44{2048}, iS45{64}]
   = T3_l_float[bS13{1}, iS15{8}, iS14{2048}, iS16{64}]
   + T1_g_float[iS3{7}, iS4{8}, iS5{2048}, iS6{64}];
(8)
T4_l_float[bS17{1}, iS18{8}, bS19{1}, iS20{2048}, iS21{64}]
   = broadcast( T3_l_float[bS13{1}, iS15{8}, iS14{2048}, iS16{64}] )
(3)
T7_l_float[bS32{1}, iS33{8}, bS34{1 ex 4}, iS35{2048}, iS36{64}] = expand( T4_l_float[bS17{1}, iS18{8}, bS19{1}, iS20{2048}, iS21{64}], {1, 8, 4, 2048, 64} )
(28)
T8_g_float[bS37{1}, iS38{8}, bS39{1 ex 4}, iS40{2048}, iS41{64}]
   = relu(T7_l_float[bS32{1}, iS33{8}, bS34{1 ex 4}, iS35{2048}, iS36{64}]);
(7)
}

} //Segmented_Fusion
```

If we pick `T9_g_float` as the reference, we'll be able to schedule the whole fusion as a single kernel.
The other output `T8_g_float` has a `bS39{1 ex 4}`, which is not mapped with `T9_g_float`, but since that iter domain doesn't participate in any transformation, it's safe to ignore it here.

Reply:
Without this part, we'll ended up with a segmentation.

I'll add a cpp test for this.

Reply:
This is fine for this particular case, but not true in general. Can you clarify why it's safe? I suppose the safety is guaranteed by some other checks, right?

Reply:
In terms of *safety*, are we talking about anything beyond the transform propagation?

I'm curious what's the counter example. i.e. when would it not be safe to have a dangling iter domain? If they didn't participate in any transform, then they shouldn't block the propagation.

For pointwise scheduler, I think the other checks [here](https://github.com/NVIDIA/Fuser/blob/76483feb4d81fec373559d318289171ac781e7b4/csrc/scheduler/pointwise.cpp#L473-L493) ensures that transformation can be propagated across the entire fusion.

Reply:
We shouldn't allow dangling IDs as they won't be scheduled. A typical symptom we have seen so far is dynamic allocation errors of register tensors.

Reply:
I see where the concern is.

Please correct me if I'm wrong, I think we'll only have dangling broadcast ids here and that should be safe.
Before this function is called to check the propagation to outputs, we run `areAllInputIdsMappedTo` against all inputs, which I think would ensure that any IDs coming into the fusion is mapped by the given reference.

So any additional IDs without a definition would only be coming from expand?!?!

Reply:
offline comment from @naoya, double check how factory methods are being handled here!

Reply:
@naoyam I think you are right about it.

https://github.com/NVIDIA/Fuser/issues/3542

So if we stay on conservative side and reject that reference, we actually would be able to schedule this.